frontend/openchat-agent/src/utils/community.ts

```diff
@@ -0,0 +1,233 @@
+            historyVisible: community.historyVisible,
+            permissions: c?.permissions ?? community.permissions,
+        };
+    });
+}
+
+export function mergeChannelUpdates(
+    channels: ChannelSummary[],
+    userCanisterUpdates: UserCanisterChannelSummaryUpdates[],
+    communityCanisterUpdates: CommunityCanisterChannelSummaryUpdates[]
+): ChannelSummary[] {
+    const userLookup = ChatMap.fromList(userCanisterUpdates);
+    const communityLookup = ChatMap.fromList(communityCanisterUpdates);
+
+    return channels.map((channel) => {
+        const u = userLookup.get(channel.id);
+        const c = communityLookup.get(channel.id);
+
+        if (u === undefined && c === undefined) return channel;
+
+        const latestMessage = c?.latestMessage ?? channel.latestMessage;
+        const readByMeUpTo = u?.readByMeUpTo ?? channel.membership.readByMeUpTo;
+
+        const blobReferenceUpdate = mapOptionUpdate(c?.avatarId, (avatarId) => ({
+            blobId: avatarId,
+            canisterId: channel.id.communityId,
+        }));
+
+        return {
+            kind: "channel",
+            id: channel.id,
+            name: c?.name ?? channel.name,
+            description: c?.description ?? channel.description,
+            minVisibleEventIndex: channel.minVisibleEventIndex,
+            minVisibleMessageIndex: channel.minVisibleMessageIndex,
+            lastUpdated: c?.lastUpdated ?? channel.lastUpdated,
+            memberCount: c?.memberCount ?? channel.memberCount,
+            public: c?.public ?? channel.public,
+            permissions: c?.permissions ?? channel.permissions,
+            historyVisible: channel.historyVisible,
+            subtype: applyOptionUpdate(channel.subtype, c?.subtype),
+            previewed: false,
+            frozen: channel.frozen, // frozen doesn't exist on CommunityCanisterChannelSummaryUpdates
+            latestEventIndex: c?.latestEventIndex ?? channel.latestEventIndex,
+            latestMessage,
+            metrics: c?.metrics ?? channel.metrics,
+            blobReference: applyOptionUpdate(channel.blobReference, blobReferenceUpdate),
+            dateLastPinned: c?.dateLastPinned ?? channel.dateLastPinned,
+            dateReadPinned: u?.dateReadPinned ?? channel.dateReadPinned,
+            gate: applyOptionUpdate(channel.gate, c?.gate) ?? { kind: "no_gate" },
+            level: "group",
+            membership: {
+                ...channel.membership,
+                mentions:
+                    c === undefined
+                        ? channel.membership.mentions
+                        : [...(c.membership?.mentions ?? []), ...channel.membership.mentions],
+                role: c?.membership?.role ?? channel.membership.role,
+                latestThreads: mergeThreads(
+                    channel.membership.latestThreads,
+                    c?.membership?.latestThreads ?? [],
+                    u?.threadsRead ?? {}
+                ),
+                readByMeUpTo:
```

Comment:
This should be `u?.readByMeUpTo ?? channel.readByMeUpTo`

Reply:
I guess it's just protecting against `readByMeUpTo` being ahead of the latest message due to a race condition.